components/bookmark/index.jsx

```diff
@@ -38,26 +39,31 @@ const useStyles = makeStyles((theme) => createStyles(styles(theme)));
 
 export const toggleBookmarkHandler = ({
   bookmarks,
-  bookmarked,
   iri,
   setSeverity,
   setMessage,
   setAlertOpen,
-  setBookmarked,
+  bookmarked,
   setBookmarks,
+  setBookmarked,
+  setDisabled,
   fetch,
 }) => {
   return async () => {
+    setDisabled(true);
+    let results;
     if (bookmarked) {
-      return; // returning for now until we have the remove logic
+      results = await removeBookmark(iri, bookmarks, fetch);
+    } else {
+      results = await addBookmark(iri, bookmarks, fetch);
     }
-    const { response, error } = await addBookmark(iri, bookmarks, fetch);
+    const { response, error } = results;
     if (error) {
       setSeverity("error");
       setMessage(error);
       setAlertOpen(true);
     } else {
-      setBookmarked(true); // optimistically set bookmarked to true so it doesn't take that long to display the item as bookmarked
+      setBookmarked(!bookmarked);
```

Comment:
To set this optimistically, should you do it above the `await` instead, and revert the change on failure?

Reply:
It is "half-optimistic" right now actually, because it doesn't wait until the bookmarks dataset is updated, just that it doesn't throw an error vs. not setting it here and having to wait both the response from the `await` and the update of the bookmarks. If I don't wait until the response is back, there is some flickering while it's waiting for the bookmarks to be updated.

Reply:
I think updating the state optimistally + not allowing further changes until updated state is resolved (i.e. disable button) might be the way to go on this?

Reply:
The button is disabled until the bookmarks are updated. The problem is that there is a brief moment where the bookmarks are updated but it hasn't checked if the bookmark already exists or not.